subprojects/docs/src/docs/userguide/dependency_management_attribute_based_matching.adoc

```diff
@@ -170,3 +170,134 @@ Because multiple values for an attribute can be _compatible_ with the requested
+<2> Inject the input artifact
+<3> Request on output location for the unzipped files
+<4> Do the actual work of the transform
+
+An artifact transform may require parameters, like a `String` determining some filter, or some file collection which is used for supporting the transformation of the input artifact.
+In order to pass those parameters to the transform action, you need to define a new type with the desired parameters.
+The type needs to implement the marker interface link:{javadocPath}/org/gradle/api/artifacts/transform/TransformParameters.html[TransformParameters].
+The properties should be <<lazy_configuration.adoc#sec:lazy_properties,read-only configurable properties>>.
+Use an interface declaring the getters and Gradle will generate the implementation.
+All getters need to have proper input annotations, see the table in the section on <<more_about_tasks.adoc#table:incremental_build_annotations,incremental build>>.
+
+Additionally to user provided parameters, an artifact transform may also be interested in the dependencies of its input artifact.
+By using the `@link:{javadocPath}/org/gradle/api/artifacts/transform/InputArtifactDependencies.html[InputArtifactDependencies]` annotation you can inject the link:{javadocPath}/org/gradle/api/file/FileCollection.html[FileCollection] which will contain the dependencies of the input artifact when the action executes.
+Note that using input artifact dependencies in a transform has performance implications, they should only be injected when required.
+
+Moreover, artifact transforms can make use of the <<build_cache.adoc#build_cache,build cache>> for caching their outputs.
+To enable the build cache for an artifact transform, add the `@link:{javadocPath}/org/gradle/api/artifacts/transform/CacheableTransform.html[CacheableTransform]` annotation on the action class.
+For a cacheable transform, it is required to annotate the `@link:{javadocPath}/org/gradle/api/artifacts/transform/InputArtifact.html[InputArtifact]` and `@link:{javadocPath}/org/gradle/api/artifacts/transform/InputArtifactDependencies.html[InputArtifactDependencies]` (when present) with normalization annotations, like e.g. `@link:{javadocPath}/org/gradle/api/tasks/PathSensitive.html[PathSensitive]`.
+
+The following example shows the above features on a transform which relocates classes.
+
+.Artifact transform for class relocation
+====
+include::sample[dir="userguide/dependencyManagement/artifactTransforms/relocate/groovy",files="build.gradle[tags=artifact-transform-relocate]"]
+include::sample[dir="userguide/dependencyManagement/artifactTransforms/relocate/kotlin",files="build.gradle.kts[tags=artifact-transform-relocate]"]
+====
+<1> Declare the transform cacheable
+<2> Interface for the transform parameters
+<3> Declare input type for each parameter
+<4> Declare a normalization for the input artifact
+<5> Inject the input artifact dependencies
+<6> Use the parameters
+
+=== Incremental Artifact Transforms
```

Comment:
What is wrong with the grammar of `Incremental Artifact Transforms`? It is in the same sense as `Incremental Tasks`.

Reply:
Sorry, it's not a problem with the grammar itself but one of consistency. I'm trying to make section headings consistent across the user manual with them focusing on something the user wants or should be doing. So something more like "Making artifact transforms incremental"

Reply:
Hmm... Does that mean that the section heading for `Incremental Tasks` should also be changed to something different? https://docs.gradle.org/release-nightly/userguide/custom_tasks.html#incremental_tasks

Reply:
Ideally, yes. I'm particularly not fond of that title because it doesn't sound any different to tasks configured for incremental build. But i dont see any need to update existing titles at this stage. That could send you down a rabbit hole.